### PR TITLE
annotate_content_models has been optimized and memory overhead reduce…

### PR DESCRIPTION
**Pull request by gwong89**
…d by 40% from original

@rtibbles  Me and @aronasorman  were wondering if the update check is necessary since it looks like all contents are not empty? line 433 in content_models.py

here  is the gist with the new improvment

https://gist.github.com/gwong89/bc11bbd727eeb715e292

### CODE REVIEW

**Comment by aronasorman** [comment]
Can you post the latest gist @gwong89?

**Comment by aronasorman** [comment]
63 MB vs. 130 MB. Nice improvement :)

File: kalite/topic_tools/content_models.py [hunk 1]
```diff
@@ -16,13 +16,14 @@
...
 import sqlite3
+import itertools
 
 from peewee import Model, SqliteDatabase, CharField, TextField, BooleanField, ForeignKeyField, PrimaryKeyField, Using, DoesNotExist, fn
 from playhouse.shortcuts import model_to_dict
 
 from .settings import CONTENT_DATABASE_PATH, CHANNEL
 from .annotate import update_content_availability
-
+from memory_profiler import profile
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: Please remove all the memory_profiler calls.

File: kalite/topic_tools/content_models.py [hunk 2]
```diff
@@ -416,19 +429,19 @@ def recurse_availability_up_tree(node, available):
...
                     for item in items:
-                        if item.kind != "Topic":
-                            item_data = unparse_model_data(model_to_dict(item, recurse=False))
-                            item_data.update(update)
-                            item_data = parse_model_data(item_data)
-                            for attr, val in item_data.iteritems():
-                                setattr(item, attr, val)
-                            item.save()
-                            recurse_availability_up_tree(item, update.get("available", False))
+                            #item_data = unparse_model_data(model_to_dict(item, recurse=False))
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: @rtibbles we removed this section as we assumed that the only thing we need to update here is the availability, not the rest of the fields.

Come to think of it, we don't have the `update` call here @gwong89.

**Comment by aronasorman** [comment]
We may need to clean this PR a bit.

File: kalite/topic_tools/content_models.py [hunk 3]
```diff
@@ -126,11 +125,8 @@ def wrapper(*args, **kwargs):
 
         if dicts and output:
             if expanded:
-                output = map(unparse_model_data, output.dicts())
-            else:
-                output = [item for item in output.dicts()]
-
-        return output
+                for dic in output:
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: We don't need to change this.

File: kalite/topic_tools/content_models.py [hunk 4]
```diff
@@ -395,11 +390,21 @@ def annotate_content_models(db=None, channel="khan", language="en", ids=None, **
     Our ids can be duplicated at the moment, so this may be several content items per id.
     When a content item has been updated, propagate availability up the topic tree.
     """
-    content_models = get_content_items(ids=ids, channel=channel, language=language)
-    updates_dict = update_content_availability(content_models)
+    def iterator_content_items(ids=None, **kwargs):
+            if ids:
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: 8 spaces. We use 4 spaces across the project.

File: kalite/topic_tools/content_models.py [hunk 5]
```diff
@@ -126,9 +125,10 @@ def wrapper(*args, **kwargs):
 
         if dicts and output:
             if expanded:
-                output = map(unparse_model_data, output.dicts())
+                for dic in output:
+                    output = map(unparse_model_data, output.dicts())
             else:
-                output = [item for item in output.dicts()]
+                    output = [item for item in output.dicts()]
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: Looks like too much indent to me.

File: kalite/topic_tools/content_models.py [hunk 6]
```diff
@@ -416,19 +425,14 @@ def recurse_availability_up_tree(node, available):
...
-            for path, update in updates_dict.iteritems():
+            for path, update in content_models:
                 if update:
                     # We have duplicates in the topic tree, make sure the stamping happens to all of them.
-                    items = Item.select().where(Item.path == path)
+                    items = Item.select().where((Item.path == path) & (Item.kind != "Topic"))
+
                     for item in items:
-                        if item.kind != "Topic":
-                            item_data = unparse_model_data(model_to_dict(item, recurse=False))
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: Deleting these lines makes this whole thing a noop. You aren't actually updating any of the content nodes.

File: kalite/topic_tools/content_models.py [hunk 7]
```diff
@@ -395,11 +394,21 @@ def annotate_content_models(db=None, channel="khan", language="en", ids=None, **
...
+    def iterator_content_items(ids=None, **kwargs):
+        if ids:
+            items = Item.select().where(Item.id.in_(ids)).dicts().iterator()
+        else:
+            items = Item.select().dicts().iterator()
 
-    with Using(db, [Item]):
+        mapped_items = itertools.imap(unparse_model_data, items)
+
+        updated_mapped_items = update_content_availability(mapped_items)
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: As mentioned verbally, we should modify ```update_content_availability``` to be a generator too, otherwise this will just cause inflation again.

**Comment by rtibbles** [comment]
I fear the main memory gains here so far are from the fact the database is not being modified at all.

File: kalite/topic_tools/annotate.py [hunk 8]
```diff
@@ -30,14 +30,15 @@ def update_content_availability(content_list, language="en"):
...
+    # turn this whole function into a generator
     try:
         contents_folder = os.listdir(django_settings.CONTENT_ROOT)
     except OSError:
         contents_folder = []
 
     subtitle_langs = {}
-
-    updates = {}
+    updates = set()
```
Comment (rtibbles) [inline kalite/topic_tools/annotate.py]: If we're just yielding an ```update``` object on every iteration, we probably don't want this here, as it will be making a reference to each yielded object and prevent them from being garbage collected.

File: kalite/topic_tools/annotate.py [hunk 9]
```diff
@@ -150,6 +148,6 @@ def update_content_availability(content_list, language="en"):
                 update["available"] = False
 
         # Path is the only unique key available.
-        updates[content.get("path")] = update
-
-    return updates
+        updates.update(content.get("id"))
```
Comment (rtibbles) [inline kalite/topic_tools/annotate.py]: See above.

File: kalite/topic_tools/content_models.py [hunk 10]
```diff
@@ -171,9 +173,9 @@ def get_content_items(ids=None, db=None, **kwargs):
     """
     with Using(db, [Item]):
         if ids:
-            values = Item.select().where(Item.id.in_(ids))
+            values = Item.select().where(Item.id.in_(ids)).iterator()
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: These changes should be reverted.

File: kalite/topic_tools/content_models.py [hunk 11]
```diff
@@ -416,19 +427,15 @@ def recurse_availability_up_tree(node, available):
                     recurse_availability_up_tree(parent, available)
 
 
-            for path, update in updates_dict.iteritems():
+            for path, update in content_models:
                 if update:
                     # We have duplicates in the topic tree, make sure the stamping happens to all of them.
-                    items = Item.select().where(Item.path == path)
+                    
+                    items = Item.select().where((Item.path == path) & (Item.kind != "Topic"))
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: Not sure why this is necessary, as path is a unique key.

File: kalite/topic_tools/content_models.py [hunk 12]
```diff
@@ -416,19 +427,15 @@ def recurse_availability_up_tree(node, available):
...
+                    Item.update("available", True).where((Item.path == path) & (Item.kind != "Topic"))
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: You are updating things irregardless of the contents of ```update``` here. It is possible that something should be flagged as unavailable.

Also, there is other data stored in update, aside from availability.

Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: See the code you have deleted for more information.

File: kalite/topic_tools/tests/annotate_tests.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+ u'{"sha": "00808cbc066910be1220235faa7d93a9126f6a57", "live": true, "id": "x1855395b96b1e34f"}',
+ u'{"sha": "90db973ee0fd319c596410e215a76a651601901d", "live": true, "id": "x477b6212a24d08da"}',
+ u'{"sha": "f6cf4d0675de30afee8cab929b97000e2c41fbc8", "live": true, "id": "xfc8946e7c80f2800"}',
+ u'{"sha": "4f522fb965ab3805f56250db187fefadb281edd2", "live": true, "id": "x14784d8f428fa949"}',
+ u'{"sha": "849c88aa2122ea5a40bb39540acf551d3ac67728", "live": true, "id": "x3f2d3b6cb53f67a4"}',
+ u'{"sha": "a9197ffb73ad014d477c7eb1eed41f9ea025774b", "live": true, "id": "xb6e923d2f396f5ab"}',
+ u'{"sha": "446eef76c02fc29049012312c704c12b4b54ba8d", "live": true, "id": "x9ef4ca7e914c87ec"}']
+    import json
+    stuff = json.dumps(thedict)
+    import pdb; pdb.set_trace()
```
Comment (rtibbles) [inline kalite/topic_tools/tests/annotate_tests.py]: This is going to cause you some issues during testing...

File: kalite/topic_tools/tests/annotate_tests.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+ u'{"sha": "bb340ca13b104a9b9d1065b74a2fdfa423f817d1", "live": true, "id": "xa756d02df7435e1a"}',
+ u'{"sha": "f65a9c4a6a8df260021fe424d562e8ff6d1741bd", "live": true, "id": "x6b9db70231ff254d"}',
+ u'{"sha": "00808cbc066910be1220235faa7d93a9126f6a57", "live": true, "id": "x1855395b96b1e34f"}',
+ u'{"sha": "90db973ee0fd319c596410e215a76a651601901d", "live": true, "id": "x477b6212a24d08da"}',
+ u'{"sha": "f6cf4d0675de30afee8cab929b97000e2c41fbc8", "live": true, "id": "xfc8946e7c80f2800"}',
+ u'{"sha": "4f522fb965ab3805f56250db187fefadb281edd2", "live": true, "id": "x14784d8f428fa949"}',
+ u'{"sha": "849c88aa2122ea5a40bb39540acf551d3ac67728", "live": true, "id": "x3f2d3b6cb53f67a4"}',
+ u'{"sha": "a9197ffb73ad014d477c7eb1eed41f9ea025774b", "live": true, "id": "xb6e923d2f396f5ab"}',
+ u'{"sha": "446eef76c02fc29049012312c704c12b4b54ba8d", "live": true, "id": "x9ef4ca7e914c87ec"}']
+    import json
```
Comment (rtibbles) [inline kalite/topic_tools/tests/annotate_tests.py]: This should be imported in the global scope.

File: kalite/topic_tools/content_models.py [hunk 15]
```diff
@@ -416,19 +427,15 @@ def recurse_availability_up_tree(node, available):
...
                     for item in items:
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: It turns out we don't need this for loop anymore; the IDs passed into this function are already updated through the `update_content_availability` call. Instead we need to make sure that each unique path is passed into the `recurse_availability_up_tree` function.

File: kalite/topic_tools/tests/annotate_tests.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    expected = {
+      "thepath" :  {'all_assessment_items': [],
+              'description': 'test',
+              'title': 'testing '}      
+    }
+    with Using(self.db, [Item]):
+      actual = update_content_availability([unparse_model_data(model_to_dict(self.item))])
+      self.assertEqual(expected, actual)
+
+  '''def test_khan_exercises_availiability_true(self):
```
Comment (aronasorman) [inline kalite/topic_tools/tests/annotate_tests.py]: Can you remove these commented out tests?

File: kalite/topic_tools/tests/annotate_tests.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+from playhouse.shortcuts import model_to_dict
+import json
+
+class AnnotateTestCase(TestCase):
+  TITLE = "testing "
+  AVAILABLE = False
+  KIND = "Exercise"
+  
+  @set_database
+  def setUp(self, db=None):
```
Comment (aronasorman) [inline kalite/topic_tools/tests/annotate_tests.py]: Your text editor keeps changing its indentation. Now you're using 2 spaces here. Before you were using 8 spaces. Can you sort this out and set this to 4 spaces?

File: kalite/topic_tools/tests/annotate_tests.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+              'title': 'testing '}      
+    }
+    with Using(self.db, [Item]):
+      actual = update_content_availability([unparse_model_data(model_to_dict(self.item))])
+      self.assertEqual(expected, actual)
+
+
+  def test_update_content_availability_false(self):
+    expected = {
+      "thepath" :  {'all_assessment_items': [],
```
Comment (aronasorman) [inline kalite/topic_tools/tests/annotate_tests.py]: The way you indent things makes things harder to read in general.

File: kalite/topic_tools/content_models.py [hunk 19]
```diff
@@ -416,19 +427,18 @@ def recurse_availability_up_tree(node, available):
...
-                        if item.kind != "Topic":
-                            item_data = unparse_model_data(model_to_dict(item, recurse=False))
-                            item_data.update(update)
-                            item_data = parse_model_data(item_data)
-                            for attr, val in item_data.iteritems():
-                                setattr(item, attr, val)
-                            item.save()
-                            recurse_availability_up_tree(item, update.get("available", False))
+                    item = Item.get().where(Item.path == path)
+                    if item.kind != "Topic":
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: @rtibbles we removed the `for` loop here thanks to your tip of `path` being unique. It also turns out that we only save about 3 MB by moving this to the DB, so you were right that we don't save much memory here.

File: kalite/topic_tools/annotate.py [hunk 20]
```diff
@@ -64,10 +64,8 @@ def update_content_availability(content_list, language="en"):
         exercise_templates = []
 
     for content in content_list:
-
         # Some nodes are duplicated, but they require the same information
         # regardless of where they appear in the topic tree
-        if content.get("id") not in updates:
```
Comment (aronasorman) [inline kalite/topic_tools/annotate.py]: @rtibbles we removed the tracking of `ids`, as per what we talked about before.

File: kalite/topic_tools/annotate.py [hunk 21]
```diff
@@ -148,8 +144,5 @@ def update_content_availability(content_list, language="en"):
             # Content is currently flagged as available, but is not. Flag as unavailable.
             if content.get("available") and "available" not in update:
                 update["available"] = False
-
-        # Path is the only unique key available.
-        updates[content.get("path")] = update
-
-    return updates
+    
+    yield content.get("path"),update
```
Comment (aronasorman) [inline kalite/topic_tools/annotate.py]: @gwong89 we normally put spaces between a `comma` and the variable.

File: kalite/topic_tools/content_models.py [hunk 22]
```diff
@@ -126,9 +127,10 @@ def wrapper(*args, **kwargs):
 
         if dicts and output:
             if expanded:
-                output = map(unparse_model_data, output.dicts())
+                for dic in output:
+                    output = map(unparse_model_data, output.dicts())
```
Comment (aronasorman) [inline kalite/topic_tools/content_models.py]: Why does this need to be in a for loop? You're doing unnecessary work here.

**Comment by aronasorman** [comment]
A lot of indentation fixes, some comments about logic.

**Comment by rtibbles** [comment]
This file still needs to be removed:
```kalite/distributed/static/django_js_reverse/js/reverse.js```

File: kalite/distributed/views.py [hunk 23]
```diff
@@ -71,7 +71,7 @@ def check_setup_status_wrapper_fn(request, *args, **kwargs):
 
 
 @render_to("distributed/learn.html")
-def learn(request):
+def learn(request, db=None):
```
Comment (rtibbles) [inline kalite/distributed/views.py]: This change is unnecessary, please remove.

File: kalite/topic_tools/annotate.py [hunk 24]
```diff
@@ -64,10 +65,8 @@ def update_content_availability(content_list, language="en"):
         exercise_templates = []
 
     for content in content_list:
-
         # Some nodes are duplicated, but they require the same information
         # regardless of where they appear in the topic tree
-        if content.get("id") not in updates:
             update = {}
```
Comment (rtibbles) [inline kalite/topic_tools/annotate.py]: Shouldn't this section get dedented then - if the 'if' statement is removed?

File: kalite/topic_tools/content_models.py [hunk 25]
```diff
@@ -395,11 +396,21 @@ def annotate_content_models(db=None, channel="khan", language="en", ids=None, **
...
+
+        mapped_items = itertools.imap(unparse_model_data, items)
+        updated_mapped_items = update_content_availability(mapped_items)
 
+        for path, update in updated_mapped_items:
+            yield path, update
+
+    with Using(db, [Item]):
+        content_models = iterator_content_items(ids=ids, channel=channel, language=language)
         with db.atomic() as transaction:
```
Comment (rtibbles) [inline kalite/topic_tools/content_models.py]: With this whole block as an atomic transaction, what are the memory implications?

**Comment by rtibbles** [comment]
These look like the right changes. One thing that needs to be removed, and I would like to see a rerun of the memory profile here too.

File: kalite/distributed/views.py (hunk 23 above)
Comment (gwong89) [inline kalite/distributed/views.py]: noted and done

File: kalite/topic_tools/annotate.py (hunk 24 above)
Comment (gwong89) [inline kalite/topic_tools/annotate.py]: noted and done

File: kalite/topic_tools/content_models.py (hunk 25 above)
Comment (gwong89) [inline kalite/topic_tools/content_models.py]: heres a gist where it does a memory profiler on the change that me and aron worked compared to the original you had on updating content. The difference is very minor

https://gist.github.com/gwong89/69e196741cab984be86a

File: kalite/topic_tools/tests/annotate_tests.py [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+            'description': 'test',
+            'title': 'testing '}      
+        }
+    
+        with Using(self.db, [Item]):
+            actual = dict(update_content_availability([unparse_model_data(model_to_dict(self.item))]))
+            self.assertEqual(expected, actual)
+
+
+    def test_update_content_availability_false(self):
```
Comment (MCGallaspy) [inline kalite/topic_tools/tests/annotate_tests.py]: @rtibbles this test and the one above are identical. Also, what are they testing in plain English? Is this a unit test for `update_content_availability`? Maybe we can spice it up a notch.
image
